mirrord-layer/src/common_hooks.rs

```diff
@@ -0,0 +1,154 @@
+use frida_gum::interceptor::Interceptor;
+
+#[cfg(target_os = "linux")]
+use crate::common_hooks::hooks::{FnDup3, FN_DUP3};
+use crate::{
+    common_hooks::hooks::{FnDup, FnDup2, FnFcntl, FN_DUP, FN_DUP2, FN_FCNTL},
+    error::LayerError,
+    replace,
+};
+
+mod ops {
+    use std::os::unix::prelude::RawFd;
+
+    use libc::c_int;
+
+    use crate::{error::LayerError, file::OPEN_FILES, socket::SOCKETS, ENABLED_FILE_OPS};
+
+    pub(super) fn fcntl(orig_fd: c_int, cmd: c_int, fcntl_fd: i32) -> Result<(), LayerError> {
+        match cmd {
+            libc::F_DUPFD | libc::F_DUPFD_CLOEXEC => dup(orig_fd, fcntl_fd),
+            _ => Ok(()),
+        }
+    }
+
+    pub(super) fn dup(fd: c_int, dup_fd: i32) -> Result<(), LayerError> {
+        if let Some(socket) = SOCKETS.lock()?.get(&fd) {
+            SOCKETS.lock()?.insert(dup_fd as RawFd, socket.clone());
+        } else if *ENABLED_FILE_OPS.get().unwrap() {
+            if let Some(file) = &&OPEN_FILES.lock()?.get(&fd) {
+                OPEN_FILES.lock()?.insert(dup_fd as RawFd, **file);
+            }
+        } else {
+            return Err(LayerError::LocalFDNotFound(fd));
+        }
+        Ok(())
+    }
+}
+
+mod hooks {
+    use libc::c_int;
+    use mirrord_macro::{hook_fn, hook_guard_fn};
+    use tracing::trace;
+
+    use super::ops::*;
+    use crate::error::LayerError;
+
+    /// https://github.com/metalbear-co/mirrord/issues/184
+    #[hook_fn]
+    pub(super) unsafe extern "C" fn fcntl_detour(fd: c_int, cmd: c_int, mut arg: ...) -> c_int {
+        trace!("fcntl_detour -> fd {:#?} | cmd {:#?}", fd, cmd);
+
+        let arg = arg.arg::<usize>();
+        let fcntl_result = FN_FCNTL(fd, cmd, arg);
+        let guard = crate::DetourGuard::new();
+        if guard.is_none() {
+            return fcntl_result;
+        }
+
+        if fcntl_result == -1 {
```

Comment:
Why not combine it with the above condition? Or even do it before, so you don't needlessly initialize a DetourGuard if fcntl_result is -1 anyway